[Fixes #61] Move must keep object reference

TFleury opened the pull request.

dnfclas commented:
Hi __@TFleury__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

Eilon commented:
@kichalla I assigned to you regarding https://github.com/aspnet/JsonPatch/issues/61.

kichalla commented on src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs:
```diff
@@ -362,6 +363,22 @@ public void Copy(Operation operation, object objectToApplyTo)
+                    }
+                }
+
                 Add(operation.path,
```
Since Add operation should either do deep copy or maintain reference based on context in which its used, we could add a flag to Add operations of adapters to either maintain reference or not. The reference check could also be moved to `ConversionResultProvider`.

kichalla commented on test/Microsoft.AspNetCore.JsonPatch.Test/NestedObjectTests.cs:
```diff
@@ -1700,6 +1700,80 @@ public void CopyToEndOfListWithSerialization()
         }
 
         [Fact]
+        public void CopyDeepCloneObject()
```
nit: Copy_DeepClonesObject

kichalla commented on test/Microsoft.AspNetCore.JsonPatch.Test/NestedObjectTests.cs:
```diff
@@ -1700,6 +1700,80 @@ public void CopyToEndOfListWithSerialization()
+            // Assert
+            Assert.Equal("C", doc.SimpleDTO.StringProperty);
+            Assert.Equal("D", doc.SimpleDTO.AnotherStringProperty);
+            Assert.Equal("C", doc.InheritedDTO.StringProperty);
```
Assert.NotSame(doc.SimpleDTO.StringProperty, doc.InheritedDTO.StringProperty);

kichalla commented on test/Microsoft.AspNetCore.JsonPatch.Test/NestedObjectTests.cs:
```diff
@@ -1753,6 +1827,77 @@ public void MoveWithSerialization()
         }
 
         [Fact]
+        public void MoveReference()
```
nit: Move_KeepsObjectReference

kichalla commented on test/Microsoft.AspNetCore.JsonPatch.Test/NestedObjectTests.cs:
```diff
@@ -1802,6 +1947,71 @@ public void MoveInListWithSerialization()
         }
 
         [Fact]
+        public void MoveReferenceInList()
```
nit: Move_KeepsObjectReferenceInList

kichalla commented on test/Microsoft.AspNetCore.JsonPatch.Test/NestedObjectTests.cs:
```diff
@@ -1802,6 +1947,71 @@ public void MoveInListWithSerialization()
+            Assert.Same(sDto1, doc.SimpleDTOList[1]);
+        }
+
+        [Fact]
```
We would need tests for the other adapters which are impacted by the changes here: Poco and Expando

TFleury commented:
@kichalla, I don't think passing a flag requesting a deep copy to the `Add` was the best solution (but it was my first idea).
`Copy` method must be responsible of making the deep copy of the original value. Doing so, we don't introduce any breaking change to `IAdapter` interface.

But my first solution was a bit ugly, and serialization/deserialization was made 2 times for a copy.

As requested, I moved type compatibility checks to `ConversionResultProvider`. I changed `ConvertTo` behavior making it trying to keep reference when possible, and added `CopyTo` method when we need to enforce deep copy.

Can you review latest commits and let me know what you think about it ?

kichalla commented:
42d0d40b36df498f6367326810852ac546c81352

Thanks! @TFleury
